### PR TITLE
dialects: (builtin) make FloatAttr a TypedAttribute

### PR DESCRIPTION
**Pull request by alexarice**
Pretty much as the title says.

### CODE REVIEW

**Comment by alexarice** [comment]
I've refactor and rebased on top of #3490

File: xdsl/parser/base_parser.py [hunk 1]
```diff
@@ -359,6 +359,36 @@ def parse_integer(
             "Expected integer literal" + context_msg,
         )
 
+    def parse_optional_float(
+        self,
+        *,
+        allow_negative: bool = False,
```
Comment (superlopuh) [inline xdsl/parser/base_parser.py:365]: I can imagine this being quite confusing, let's drop the default argument for now?

Comment (alexarice) [reply xdsl/parser/base_parser.py:365]: It was copied off the above function for integers I think, but sounds like a good idea

Comment (alexarice) [reply xdsl/parser/base_parser.py:365]: It seems I copied poorly and the rest have a default value of True, would that be preferred or still no default?

Comment (superlopuh) [reply xdsl/parser/base_parser.py:365]: I have a slight preference towards default True

File: xdsl/parser/base_parser.py [hunk 2]
```diff
@@ -376,8 +406,7 @@ def parse_optional_number(
         ) is not None:
             return -value if is_negative else value
 
-        if (value := self._parse_optional_token(Token.Kind.FLOAT_LIT)) is not None:
-            value = value.get_float_value()
+        if (value := self.parse_optional_float()) is not None:
```
Comment (superlopuh) [inline xdsl/parser/base_parser.py:409]: ?
```suggestion
        if (value := self.parse_optional_float(allow_negative=False)) is not None:
```

Comment (alexarice) [reply xdsl/parser/base_parser.py]: ? indeed

**Review by math-fehr — APPROVED** [review]
Nice!

**Pull request merged**